include/nighthawk/adaptive_load/adaptive_load_controller.h

```diff
@@ -0,0 +1,39 @@
+#pragma once
+
+#include "envoy/common/time.h"
+
+#include "api/adaptive_load/adaptive_load.pb.h"
+#include "api/client/service.grpc.pb.h"
+
+namespace Nighthawk {
+namespace AdaptiveLoad {
+
+/**
+ * Performs an adaptive load session consisting of the Adjusting Stage and the
+ * Testing Stage. Adjusting Stage: Runs a series of short benchmarks, checks metrics according to
+ * MetricSpecs, and adjusts load up or down based on the result; returns an error if convergence is
+ * not detected before the deadline in the spec. Load adjustments and convergence detection are
+ * computed by a StepController plugin. Metric values are obtained through MetricsPlugins. Testing
+ * Stage: When the optimal load is found, runs one long benchmark to validate it.
+ *
+ * @param nighthawk_service_stub A Nighthawk Service gRPC stub.
+ * @param spec A proto that defines all aspects of the adaptive load session, including metrics,
+ * threshold, duration of adjusting stage benchmarks, and underlying Nighthawk traffic parameters.
+ * @param diagnostic_ostream A place to write progress messages, such as std::cerr to write to the
+ * console, a std::ostringstream to store the message in memory, or a std::ostream adapter around a
+ * custom logging system.
+ * @param time_source An abstraction of the system clock. Normally, just construct an
+ * Envoy::Event::RealTimeSystem and pass it. NO_CHECK_FORMAT(real_time). If calling from an
+ * Envoy-based process, there may be an existing TimeSource or TimeSystem to use. If calling
+ * from a test, pass a fake TimeSource.
+ *
+ * @return AdaptiveLoadSessionOutput a proto logging the result of all traffic attempted and all
+ * corresponding metric values and scores.
+ */
```

Comment:
Cleaned up the implementation to throw no exceptions, documented the fact that all errors are in the output proto.

I was reluctant to declare "noexcept" because it would spread like a virus, but theoretically we could.

Reply:
Agreed, we shouldn't declare noexcept. Just documenting it is enough.